### PR TITLE
[Accessibility] Fixing ComboBoxAccessibleObject issue when Narrator and Inspect don't respond

### LINKED ISSUE #3452: [Accessibility] ComboBox control causes Narrator and Inspect  to stop responding
* .NET Core Version: 
.Net 5.0.100-preview.7.20314.1 from 5.0 master branch

* Have you experienced this same bug with .NET Framework?: 
No

**More Info:**

1. This issue can also reproduce on without Private.
2. This issue can reproduce in RTL and LTR environment.

**Minimal repro:**

1. Open the attached ‘WindowsFormsApp3’ application in VS, then build and run it.
2. Open Narrator or Inspect, then focus it's cursor to the ComboBox control.
[WindowsFormsApp4.zip](https://github.com/dotnet/winforms/files/4785850/WindowsFormsApp4.zip)

**Expected behavior:**
Narrator or Inspect's cursor focuses on the ComboBox control and works fine.

**Actual behavior:**
ComboBox control causes Narrator and Inspect not respond.
1909-NewIssue2

### PR DESCRIPTION
**Pull request by vladimir-krestov**
Fixes #3452

## Proposed changes

- Just remove `[ComImport]` attribute

## Customer Impact

- ComboBox doesn't break Narrator and Inspect

## Regression? 

- Yes, from 423b2bc5726b4109741f189e34e5f37f90140d01

## Risk

- Minimal

### Before
- Inspect doesn't response after selecting ComboBox editing control
bug

### After
- ComboBox AccessibleObject works well
Fixed

## Test methodology 

- Manual testing 
- CTI

## Accessibility testing  
- Using Narrator and Inspect
 

## Test environment(s) 
- Microsoft Windows [Version 10.0.18363.900]
- .Net Version: 5.0.0-preview.8.20326.9

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3524)

### CODE REVIEW

**Comment by vladimir-krestov** [comment]
This bug was added when https://github.com/dotnet/winforms/commit/423b2bc5726b4109741f189e34e5f37f90140d01 was merged.
The bug is not reproduced if to remove `[ComImport]` attribute.

@hughbe, @JeremyKuhne 
Why do we need this attribute?
Why removing the attribute fixes the issue of ComboBox?

**Comment by hughbe** [comment]
I’m sorry I don’t know if this right? Can’t understand why removing ComImport would fix things... this is a com import

**Comment by JeremyKuhne** [comment]
@AaronRobinsonMSFT is there perhaps something funky with putting `[ComImport]` on both a base _and_ a derived interface?

``` C#
        [ComImport]
        [Guid("D6DD68D1-86FD-4332-8666-9ABEDEA2D24C")]
        [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
        public interface IRawElementProviderSimple {}

        [ComImport] // <-- broke things??
        [Guid("1d5df27c-8947-4425-b8d9-79787bb460b8")]
        [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
        public interface IRawElementProviderHwndOverride : IRawElementProviderSimple {}
```

**Comment by AaronRobinsonMSFT** [comment]
>something funky with putting [ComImport] on both a base and a derived interface?

Not that I know of. However, `IRawElementProviderHwndOverride` doesn't inherit from `IRawElementProviderSimple` - see [IRawElementProviderHwndOverride](https://docs.microsoft.com/windows/win32/api/uiautomationcore/nn-uiautomationcore-irawelementproviderhwndoverride#inheritance). This is also observed in the IDL:

``` C++
MIDL_INTERFACE("1d5df27c-8947-4425-b8d9-79787bb460b8")
IRawElementProviderHwndOverride : public IUnknown
{
public:
    virtual HRESULT STDMETHODCALLTYPE GetOverrideProviderForHwnd( 
        /* [in] */ __RPC__in HWND hwnd,
        /* [retval][out] */ __RPC__deref_out_opt IRawElementProviderSimple **pRetVal) = 0;
    
};
```

Even if it was, the inheriting interface needs to be declared in the derived type. Can't find a good link, but in general it would be something like:

```CSharp
interface IFoo
{
   void Bar();
}

interface IFoo2 : IFoo
{
    new void Bar();

    void Baz();
}
```

**Comment by AaronRobinsonMSFT** [comment]
Sigh... I just looked at [`IRawElementProviderHwndOverride`](https://docs.microsoft.com/dotnet/api/system.windows.automation.provider.irawelementproviderhwndoverride). That isn't correct, but I don't think it would be easy to change now. I don't see why this would impact the system at all, but I could see it messing with the runtime if the type is defined to inherit in someone way and it doesn't actually.

**Comment by hughbe** [comment]
Cref https://github.com/dotnet/winforms/pull/3231 which should fux

**Comment by AaronRobinsonMSFT** [comment]
@hughbe We need to be careful about changing this type. This is a breaking change from the .NET side of things even if it is wrong with respect to the actual COM types. @RussKie and @JeremyKuhne What are your thoughts on this breaking change?

I have no opinion either way. Let's just make sure we understand the user impact.

**Comment by JeremyKuhne** [comment]
> However, IRawElementProviderHwndOverride doesn't inherit from IRawElementProviderSimple

Ugh, sorry I missed this. :)

> What are your thoughts on this breaking change?

As we've documented it as deriving we should probably leave it that way. We can just comment fully as to the weirdness with this particular instance in the code so we handle it more carefully in the future.

**Comment by RussKie** [comment]
If the accessibility tools continue to function (such as Narrator and Accessibility Insights) I think we can take the change.

/cc: @merriemcgaw

**Comment by hughbe** [comment]
> @hughbe We need to be careful about changing this type. This is a breaking change from the .NET side of things even if it is wrong with respect to the actual COM types. @RussKie and @JeremyKuhne What are your thoughts on this breaking change?

Is this a breaking change? AFAICT it’s an internal interop type

**Comment by RussKie** [comment]
Yes, you're right - all these types live in `internal static partial class UiaCore`, so they are not directly exposed to our consumers, only indirectly via `Control.AccessibilityObject` properties IIUC.

**Comment by AaronRobinsonMSFT** [comment]
If that is true, where is the type described by the referenced documentation above?

**Comment by RussKie** [comment]
> If that is true, where is the type described by the referenced documentation above?

I believe these docs are from WPF: https://github.com/dotnet/wpf/blob/ae1790531c3b993b56eba8b1f0dd395a3ed7de75/src/Microsoft.DotNet.Wpf/src/UIAutomation/UIAutomationProvider/System/Windows/Automation/Provider/IRawElementProviderHwndOverride.cs

/cc: @fabiant3 @ryalanms

**Comment by AaronRobinsonMSFT** [comment]
>I believe these docs are from WPF:

Ah. Okay. Then this seems like changing the definition for WinForms is appropriate since it is an internal implementation detail.

**Comment by vladimir-krestov** [comment]
Testers approved these changes ✔️
Inspect crashes without the fix and works well if to remove `[ComImport]`

File: src/System.Windows.Forms.Primitives/src/Interop/UiaCore/Interop.IRawElementProviderHwndOverride.cs [hunk 1]
```diff
@@ -13,7 +13,6 @@ internal static partial class UiaCore
         ///  Implemented by providers which want to provide information about or want to
         ///  reposition contained HWND-based elements.
         /// </summary>
-        [ComImport]
         [Guid("1d5df27c-8947-4425-b8d9-79787bb460b8")]
         [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
```
Comment (weltkante) [inline src/System.Windows.Forms.Primitives/src/Interop/UiaCore/Interop.IRawElementProviderHwndOverride.cs:17]: @RussKie @JeremyKuhne 

I don't think this is right, if ComImport is removed the others should also be removed for consistency, otherwise you risk the dotnet runtime doing the wrong thing (you don't own the interface declaration, so it must be import or nothing)

Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/src/Interop/UiaCore/Interop.IRawElementProviderHwndOverride.cs:17]: I'm not sure I follow. `[ComImport]` itself doesn't expose it, it just generates a RCW/CCW if you pass the implementing class through interop.

From what I read we should remove the derivation, not the `[ComImport]` if we're exposing the implementing objects in any way to native. Or am I missing something?

Comment (weltkante) [reply src/System.Windows.Forms.Primitives/src/Interop/UiaCore/Interop.IRawElementProviderHwndOverride.cs:17]: I'm generally just uncomfortable with removing `[ComImport]` without removing the GUID referring to an interface you don't own, considering the default if no `[ComImport]` is given would be an implicit `[ComVisible(true)]` meaning you claim that you own the interface.

That should not happen since we specified the assembly attribute but personally I don't understand all the code paths, a lot of logic just defaults to assuming non-import interfaces means the assembly owns the declaration. Any such decision must be avoided since it would be wrong.

To my understanding with this PR the `[Guid]`/`[InterfaceType]` attributes are now no longer having any effect so can and should be removed together with `[ComImport]`. If I'm missing something and the other two attributes are still doing something I'd like to know what happens, because I suspect it might just superficially work but might be wrong when looking at it in detail.

TLDR: to me it looks like this is now a plain managed interface and not performing any interop purpose besides forcing a base interface implementation, so the other two attributes should be removed together with `ComImport` to prevent mistaking the interface for something it isn't

**Comment by Tanya-Solyanik** [comment]
I'm not even sure that this fix is working properly, I can't click around items in the combobox drop downs in our test application when inspect is running, at some point inspect hangs.

**Comment by merriemcgaw** [comment]
@RussKie to keep an eye on this one too.

**Comment by Tanya-Solyanik** [comment]
@JeremyKuhne, @AaronRobinsonMSFT -
>From what I read we should remove the derivation, not the [ComImport] if we're exposing the implementing objects in any way to native. Or am I missing something?

I experimented like this:
1. removed derivation and kept [ComImport], => inspect does not work
2. keep derivation+ remove [ComImport] => inspect works
3. remove derivation and remove [ComImport] => inspect works

These results don't make sense to me. I think we are covering up a bigger issue here. Do you have any suggestions?

**Comment by AaronRobinsonMSFT** [comment]
@Tanya-Solyanik @hughbe @JeremyKuhne I truthfully I am unclear on what is going on here. All I noted was the type definition isn't correct. Can someone create a minimal example that I can debug?

**Comment by vladimir-krestov** [comment]
@AaronRobinsonMSFT, I use the internal test forms in our project. ../tests/integration/WinformsControlsTest

It's comfortable to make changes in the code and just run the test project without moving any dlls.
Run the test project and click ComboBoxes. You can use it as a minimal example that you can debug.

image

**Comment by AaronRobinsonMSFT** [comment]
@vladimir-krestov Unfortunately I can't really use Narrator so am having a tough time reproducing this issue. I am on Remote Desktop from a mac and that doesn't seem to work with Narrator. Any way we could reduce the issue to some other approach? I will try a few other things as well.

**Comment by vladimir-krestov** [comment]
@AaronRobinsonMSFT, can you use Inspect?

**Comment by AaronRobinsonMSFT** [comment]
@vladimir-krestov I stand corrected. I am able to use Narrator - was a setting on the macOS RDP client. On the above example, I am able to hear it say "ComboBox". Can we elaborate on what I should or shouldn't hear?

**Comment by vladimir-krestov** [comment]
@AaronRobinsonMSFT, Narrator and Inspect don't respond if to click to an edit field of a ComboBox.
After deleting [ComImport] attribute everything works well. If to return this attribute your Inspect will crash. 
And we don't know is my fix correct? Or it is a latent bigger issue.

**Comment by AaronRobinsonMSFT** [comment]
I am also able to hear the Narrator enumerate each element in the ComboBox when expanded as well as state the count.

**Comment by AaronRobinsonMSFT** [comment]
This isn't working at all. In fact by removing the `ComImport` the `IRawElementProviderHwndOverride` type is never queried by Narrator and therefore `GetOverrideProviderForHwnd` is never called - which is why it "works". This is an area I am very unfamiliar with but there is something wrong with how `IRawElementProviderHwndOverride` is used and it's implementation is a bit convoluted.

Someone from WinForms or with far more experience/knowledge of the Accessibility API needs to actually debug the implementation instead of just making sure it doesn't crash. I can help out, but this doesn't seem like a specific COM issue since the interface itself appears very straight forward. If someone can point to a specific COM interop issue I am more than willing to investigate but the current issue is simply using this Accessibility interface in general.
